File: front/lib/domain/model/basket_exchange.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'basket_exchange.freezed.dart';
part 'basket_exchange.g.dart';

/// Wire status for a [BasketExchange] offer.
///
/// Mirrors `BasketExchangeStatus` on the back. Uppercase wire values.
enum BasketExchangeStatus {
  @JsonValue('OPEN')
  open,
  @JsonValue('ACCEPTED')
  accepted,
  @JsonValue('CANCELLED')
  cancelled,
}

/// Wire status for a [BasketExchangeRequest] embedded in a [BasketExchange].
///
/// Mirrors `BasketExchangeRequestStatus` on the back. Uppercase wire values.
enum BasketExchangeRequestStatus {
  @JsonValue('PENDING')
  pending,
  @JsonValue('ACCEPTED')
  accepted,
  @JsonValue('REJECTED')
  rejected,
  @JsonValue('WITHDRAWN')
  withdrawn,
}

/// A request from one member to take over another member's basket slot.
///
/// Embedded inside [BasketExchange.requests]. The back allocates real request
/// ids server-side: when the client submits a `tmp_*` [requestId] inside
/// `BasketExchange.requests[]`, the [MutationOutcome.serverEntityId] carries
/// the outer [BasketExchange.basketExchangeId] (not the request id). The front
/// recovers the allocated request id by re-reading the response [BasketExchange]
/// after sync — the `tmp_*` request entry is replaced with the real one.
@freezed
abstract class BasketExchangeRequest with _$BasketExchangeRequest {
  const factory BasketExchangeRequest({
    @JsonKey(name: 'request_id') required String requestId,
    @JsonKey(name: 'requester_member_id') required String requesterMemberId,
    // ISO-8601 instant string, e.g. "2026-05-26T12:00:00Z".
    @JsonKey(name: 'created_at') required String createdAt,
    required BasketExchangeRequestStatus status,
    // ISO-8601 instant string; null/absent when not yet decided.
    @JsonKey(name: 'decided_at') String? decidedAt,
    // Reciprocal swap: the delivery (and optional contract) the requester offers
    // in return. The offerer receives this basket when validating the request.
    // Required at submission time (enforced by the back); nullable on the wire for
    // robustness / legacy rows.
    @JsonKey(name: 'proposed_delivery_id') String? proposedDeliveryId,
    @JsonKey(name: 'proposed_contract_id') String? proposedContractId,
  }) = _BasketExchangeRequest;

  factory BasketExchangeRequest.fromJson(Map<String, Object?> json) =>
      _$BasketExchangeRequestFromJson(json);
}

/// A basket-exchange offer published by a member on one of their upcoming
/// delivery slots.
///
/// Synced on the `organization:{id}` scope. The back never allows
/// [applyDelete] — tombstones are [FORBIDDEN].
///
/// `tmp_*` ids: when the client creates a new offer, it assigns a `tmp_*`
/// [basketExchangeId]. After sync the handler reads [MutationOutcome.serverEntityId]
/// to remap the local row to the real id. For embedded [requests], see
/// [BasketExchangeRequest] doc comment for the id-recovery convention.
@freezed
abstract class BasketExchange with _$BasketExchange {
  const factory BasketExchange({
    @JsonKey(name: 'basket_exchange_id') required String basketExchangeId,
    @JsonKey(name: 'organization_id') required String organizationId,
    @JsonKey(name: 'delivery_id') required String deliveryId,
    @JsonKey(name: 'contract_id') required String contractId,
    @JsonKey(name: 'offering_member_id') required String offeringMemberId,
    String? motive,
    required BasketExchangeStatus status,
    // ISO-8601 instant string, e.g. "2026-05-26T12:00:00Z".
    @JsonKey(name: 'created_at') required String createdAt,
    // ISO-8601 instant string; null/absent until the exchange is decided.
    @JsonKey(name: 'decided_at') String? decidedAt,
    // Null/absent while no request has been accepted yet.
    @JsonKey(name: 'accepted_request_id') String? acceptedRequestId,
    @Default([]) List<BasketExchangeRequest> requests,
  }) = _BasketExchange;

  factory BasketExchange.fromJson(Map<String, Object?> json) =>
      _$BasketExchangeFromJson(json);
}

File: back/persistence/model/src/main/kotlin/BasketExchange.kt
@file:OptIn(ExperimentalTime::class)

package persistence.model

import id.Id
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.ExperimentalTime
import kotlin.time.Instant

/**
 * A basket-exchange offer published by a member who cannot attend a delivery.
 * Requests are embedded in the offer as an aggregate root (1 mutation = 1 transaction).
 *
 * Scope: organization:{organizationId}
 */
@Serializable
data class BasketExchange(
    @SerialName("basket_exchange_id") val basketExchangeId: Id<BasketExchange>,
    @SerialName("organization_id") val organizationId: Id<Organization>,
    @SerialName("delivery_id") val deliveryId: Id<Delivery>,
    @SerialName("contract_id") val contractId: Id<DeliveryContract>,
    @SerialName("offering_member_id") val offeringMemberId: Id<Member>,
    val motive: String? = null,
    val status: BasketExchangeStatus,
    @SerialName("created_at") val createdAt: Instant,
    @SerialName("decided_at") val decidedAt: Instant? = null,
    @SerialName("accepted_request_id") val acceptedRequestId: Id<BasketExchangeRequest>? = null,
    val requests: List<BasketExchangeRequest> = emptyList(),
)

@Serializable
enum class BasketExchangeStatus {
    OPEN,
    ACCEPTED,
    CANCELLED,
}

/**
 * A request sent by a member who wants to take a basket-exchange offer.
 * Embedded inside [BasketExchange.requests].
 *
 * Reciprocal swap: the requester proposes one of their own deliveries
 * ([proposedDeliveryId] / [proposedContractId]) in return — the offerer receives
 * that basket when validating the request. The fields are nullable on the wire
 * (robustness / legacy rows) but required at submission time by the service.
 */
@Serializable
data class BasketExchangeRequest(
    @SerialName("request_id") val requestId: Id<BasketExchangeRequest>,
    @SerialName("requester_member_id") val requesterMemberId: Id<Member>,
    @SerialName("created_at") val createdAt: Instant,
    val status: BasketExchangeRequestStatus,
    @SerialName("decided_at") val decidedAt: Instant? = null,
    @SerialName("proposed_delivery_id") val proposedDeliveryId: Id<Delivery>? = null,
    @SerialName("proposed_contract_id") val proposedContractId: Id<DeliveryContract>? = null,
)

@Serializable
enum class BasketExchangeRequestStatus {
    PENDING,
    ACCEPTED,
    REJECTED,
    WITHDRAWN,
}
